ops/pebble.py

```diff
@@ -308,29 +317,97 @@ def __repr__(self):
                 ).format(self=self)
 
 
-class API:
+class Setup:
+    """Represents a Pebble setup layer (or flattened setup).
+
+    This is not documented, but captured in code here:
+    https://github.com/canonical/pebble/blob/master/internal/setup/setup.go
+    """
+
+    def __init__(self, raw: Union[str, Dict] = None):
+        if isinstance(raw, str):
+            d = yaml.safe_load(raw) or {}
+        else:
+            d = raw or {}
+        self.summary = d.get('summary', '')
+        self.description = d.get('description', '')
+        self.services = {name: SetupService(name, service)
+                         for name, service in d.get('services', {}).items()}
+
+    def to_yaml(self) -> str:
+        """Convert this setup to its YAML representation."""
+        return yaml.dump(self.to_dict(), Dumper=_safe_dumper)
+
+    def to_dict(self) -> Dict:
+        """Convert this setup to a dict representation."""
+        fields = [
+            ('summary', self.summary),
+            ('description', self.description),
+            ('services', {name: service.to_dict() for name, service in self.services.items()})
+        ]
+        return {name: value for name, value in fields if value}
+
+    def __repr__(self) -> str:
+        return 'Setup({!r})'.format(self.to_dict())
+
+    __str__ = to_yaml
+
+
+class SetupService:
+    """Represents a service description in a Pebble setup layer."""
+
+    def __init__(self, name: str, raw: Dict = None):
+        self.name = name
+        raw = raw or {}
+        self.summary = raw.get('summary', '')
+        self.description = raw.get('description', '')
+        self.default = raw.get('default', '')
+        self.override = raw.get('override', '')
+        self.command = raw.get('command', '')
+        self.after = raw.get('after', [])
+        self.before = raw.get('before', [])
+        self.requires = raw.get('requires', [])
+        self.environment = raw.get('environment') or {}
+
+    def to_dict(self) -> Dict:
+        """Convert this service object to its dict representation."""
+        fields = [
+            ('summary', self.summary),
+            ('description', self.description),
+            ('default', self.default),
+            ('override', self.override),
+            ('command', self.command),
+            ('after', self.after),
+            ('before', self.before),
+            ('requires', self.requires),
+            ('environment', self.environment),
+        ]
+        return {name: value for name, value in fields if value}
+
+    def __repr__(self) -> str:
+        return 'SetupService({!r})'.format(self.to_dict())
+
+
+class PebbleClient:
     """Pebble API client."""
 
     def __init__(self, socket_path=None, opener=None, base_url='http://localhost', timeout=5.0):
-        """Initialize an instance. Defaults to Unix socket "$PEBBLE/.pebble.socket"."""
```

Comment:
Perhaps something resembling:
```
"""Initialize client instance.

Arguments:

    socket_path: .... - defaults to $PEBBLE/.pebble.socket.

...
"""
```

Reply:
I think you're looking at the deleted code -- I've actually removed the environment variable default and the env var handling from the `PebbleClient` class, and only use that in the test command line client. In `model.py` the socket_path arg is always passed in.

Reply:
Duh, thanks.